lib/api/v3/work_packages/work_package_representer.rb

```diff
@@ -593,6 +593,20 @@ def spent_time=(value)
           # noop
         end
 
+        def duration=(value)
+          unless OpenProject::FeatureDecisions.work_packages_duration_field_active?
+            raise NoMethodError,
+                  "#{WorkPackage.human_attribute_name(:duration)} #{I18n.t('activerecord.errors.messages.does_not_exist')}"
+          end
+          if represented.milestone?
+            raise NoMethodError, I18n.t('activerecord.errors.messages.duration_for_milestones_not_writable')
+          end
```

Comment:
That part belongs into the contract.

With the `if` statement the way it currently is (see above), I would also have assumed that this case here is never reached. 

Within the contract, we could either use the `writeable` property to turn it non writeable which will result in a standard "is not writeable" error message. But in this case I think what you did, having a custom error makes sense. The only thing I would change about the message is the wording, saying something more like: 

"Milestone typed work packages don't have a duration."

Reply:
Thank you for the thorough explanation. I updated the error message: 0bb487ab269afa8c4b0420e88e4b46e072dd348a

Reply:
Thanks for the update of the message.

When testing this, I noticed, that no error message is thrown  when trying it for a milestone typed work package. The reason is, that we use a [`ParserStruct`](https://github.com/opf/openproject/blob/dev/app/services/api/parser_struct.rb) to collect the passed in data. This is done to avoid side effects, e.g. assigning to an `has_many` relationship would directly persist the change. As a `ParserStruct` is used, it will always respond with `nil` to `represented.milestone?` Because of that, and also because all domain validations should be placed in the contract, please place that check in the [`WorkPackages::BaseContract`](https://github.com/opf/openproject/blob/dev/app/contracts/work_packages/base_contract.rb#L147-L148)

Reply:
Done: 1ff8990f9ac3037fa4405c630e49d9492be8343b